sktime/clustering/_k_means.py

```diff
@@ -0,0 +1,174 @@
+# -*- coding: utf-8 -*-
+"""Time Series K-Means Clusterer."""
+
+__author__ = ["Christopher Holder", "Tony Bagnall"]
+__all__ = ["TimeSeriesKMeans"]
+
+from sktime.clustering.base._typing import (
+    InitAlgo,
+    AveragingAlgo,
+    AveragingAlgoDict,
+    MetricParameter,
+    NumpyArray,
+    NumpyOrDF,
+    NumpyRandomState,
+)
+from sktime.clustering.partitioning._averaging_metrics import (
+    BarycenterAveraging,
+    MeanAveraging,
+)
+from sktime.clustering.partitioning._lloyds_partitioning import (
+    TimeSeriesLloydsPartitioning,
+)
+
+
+class TimeSeriesKMeans(TimeSeriesLloydsPartitioning):
+    """Time Series K-Means Clusterer
+
+    Parameters
+    ----------
+    n_clusters: int, default = 8
+        The number of clusters to form as the number of
+        centroids to generate.
+
+    init_algorithm: Init_Algo, default = forgy
+        Algorithm that is used to initialise the cluster
+        centers. str options are "forgy", "random" or
+        "k-means++". If using custom center init algorithm
+        then must be of type Init_Algo
+
+    max_iter: int, default = 300
+        Maximum number of iterations of time series k means
+        for a single run.
+
+    verbose: bool, default = False
+        Verbosity mode.
+
+    metric: Metric_Parameter, default = None
+        The distance metric that is used to calculate the
+        distance between points.
+
+    averaging_algorithm: Averaging_Algo, default = "mean"
+        The method used to create the average from a cluster.
+        str options are "dba" dtw barycenter averaging and
+        "means" for mean average.
+
+    averaging_algorithm_iterations: int, default = 10
+        Where appropriate (i.e. DBA) the average is refined by
+        iterations. This is the number of times it is refined
+
+    random_state: NumpyRandomState, default = np.random.RandomState(1)
+        Generator used to initialise the centers.
+    """
+
+    __averaging_algorithm_dict: AveragingAlgoDict = {
+        "dba": BarycenterAveraging,
+        "mean": MeanAveraging,
+    }
+
+    def __init__(
+        self,
+        n_clusters: int = 8,
+        init_algorithm: InitAlgo = "forgy",
+        max_iter: int = 300,
+        verbose: bool = False,
+        metric: MetricParameter = "dtw",
+        averaging_algorithm: AveragingAlgo = "mean",
+        averaging_algorithm_iterations: int = 10,
+        random_state: NumpyRandomState = None,
+    ):
+        super(TimeSeriesKMeans, self).__init__(
+            n_clusters=n_clusters,
+            init_algorithm=init_algorithm,
+            max_iter=max_iter,
+            verbose=verbose,
+            metric=metric,
+            random_state=random_state,
+        )
+        self.averaging_algorithm = averaging_algorithm
+        self.averaging_algorithm_iterations = averaging_algorithm_iterations
+        self._averaging_algorithm = None
+
+    def fit(self, X: NumpyOrDF) -> None:
```

Comment:
We may want to consider in the future to add `y` which is always ignored to be able to reuse composition functionality from classification (e.g. pipelines or tuning). 
```suggestion
    def fit(self, X: NumpyOrDF, y : NumpyArray = None) -> None:
```